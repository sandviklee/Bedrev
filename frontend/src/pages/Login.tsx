import { Helmet } from "react-helmet";
import Field from "../components/Input/Field/Field";
import * as Form from "@radix-ui/react-form";
import Button from "../components/Buttons/Button";
import { Link, useNavigate } from "react-router-dom";
import bedrevIcon from "/bedrevIcon.svg";
import bedrevLogoWhite from "/bedrevLogoWhite.svg";
import homepageArt from "/homepageArt.svg";
import bedrevLogo from "/bedrevLogo.svg";
import { gql, useMutation } from "@apollo/client";
import { useState } from "react";

const Login = () => {
  const [emailValue, setEmailValue] = useState("");
  const [passwordValue, setPasswordValue] = useState("");
  const [errorValue, setErrorValue] = useState<string>("");
  const navigate = useNavigate();

  const SIGNUP_MUTATION = gql`
        mutation {
        login(email: "${emailValue}", password: "${passwordValue}") {
            token
            }
        }`;

  const [login] = useMutation(SIGNUP_MUTATION, {
    onCompleted: async ({ login }) => {
      setErrorValue("");
      await localStorage.setItem("auth-token", login.token);
      navigate(-1); //Navigate to previous page.
    },
    onError: (error) => {
      setErrorValue(error.message);
    },
  });

  return (
    <>
      <Helmet>
        <meta charSet="utf-8" />
        <title>bedrev | Login</title>
        <link rel="icon" type="image/svg+xml" href={bedrevIcon} />
      </Helmet>
      <div className="w-full flex flex-row font-whyte">
        <div className="bg-purple w-[0px] h-screen sm:w-1/3 sm:min-w-[200px] sm:p-10 sm:pt-[8vh]">
          <Link to="/">
            <img
              src={bedrevLogoWhite}
              alt="companylogo"
              className="mb-[12vh]"
            />
          </Link>

          <img src={homepageArt} alt="login" />
        </div>
        <div className="flex-1 flex flex-col items-center font-whyte pt-[100px] sm:pt-[25vh] mx-[50px] sm:mx-[0px]">
          <div className="w-full sm:min-w-[400px] sm:w-1/2">
            <Link to="/">
              <img
                src={bedrevLogo}
                alt="companylogo"
                className="mb-8 sm:hidden"
              />
            </Link>
            <h3 className="text-3xl text-primary tracking-wider font-light">
              Velkommen
            </h3>
            <p className="pt-[20px] text-primary">
              Fyll inn innloggingsdetaljene dine for å logge inn
            </p>
            <Form.Root
              className="w-full mt-[20px] flex flex-col gap-4"
              onSubmit={(event) => {
                event.preventDefault();
              }}
            >
              <Field
                name="email"
                label="E-post"
                type="text"
                border={true}
                direction="col"
                placeholder="Skriv inn din E-post"
                value={emailValue}
                setValue={setEmailValue}
              />
              <Field
                name="password"
                label="Passord"
                type="password"
                border={true}
                direction="col"
                placeholder="Skriv inn ditt passord"
                value={passwordValue}
                setValue={setPasswordValue}
              />
              <div className="mt-2 w-full">
                <Button text="Logg inn" type="submit" onClick={() => login()} />
              </div>
              {errorValue !== "" && (
                <p className="text-red tracking-wide">{errorValue}</p>
              )}
              <Link to="/registrer" className="text-yellow self-end mt-1">
                Ingen bruker fra før? Registrer deg her
              </Link>
            </Form.Root>
          </div>
        </div>
      </div>
    </>
  );
};

export default Login;
