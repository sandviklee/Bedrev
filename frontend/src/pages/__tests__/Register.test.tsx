import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { MockedProvider } from "@apollo/client/testing";
import { gql } from "@apollo/client";
import Login from "../Registrer.tsx";

describe("Register page", () => {
  const SIGNUP_MUTATION = gql`
    mutation Signup($email: String!, $password: String!) {
      signup(email: $email, password: $password) {
        token
      }
    }
  `;

  const mocks = [
    {
      request: {
        query: SIGNUP_MUTATION,
        variables: {
          email: "test@test.com",
          password: "testpassword",
        },
      },
      result: {
        data: {
          signup: {
            token: "testtoken",
          },
        },
      },
    },
  ];

  test("renders correctly", () => {
    render(
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Login />
        </MockedProvider>
      </Router>,
    );

    expect(screen.getByText("Registrer ny bruker")).toBeInTheDocument();
    expect(
      screen.getByText("Fyll inn feltene for Ã¥ registrere deg"),
    ).toBeInTheDocument();
    expect(screen.getByText("Registrer deg")).toBeInTheDocument();
    expect(
      screen.getByText("Har du allerede bruker? Logg inn her"),
    ).toBeInTheDocument();

    // check if fields are there
    expect(
      screen.getByPlaceholderText("Skriv inn din E-post"),
    ).toBeInTheDocument();
    const passwordFields = screen.getAllByPlaceholderText(
      "Skriv inn ditt passord",
    );
    expect(passwordFields.length).toBe(2);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Login />
        </MockedProvider>
      </Router>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("navigate to correct route", () => {
    render(
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Login />
        </MockedProvider>
      </Router>,
    );
    const button = screen
      .getByText("Har du allerede bruker? Logg inn her")
      .closest("a");
    expect(button?.getAttribute("href")).toBe("/logg-inn");
  });
});
