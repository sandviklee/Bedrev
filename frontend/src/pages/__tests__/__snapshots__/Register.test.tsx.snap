// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register page > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full flex flex-row font-whyte"
  >
    <div
      class="bg-purple w-[0px] h-screen sm:w-1/3 sm:min-w-[200px] sm:p-10 sm:pt-[8vh]"
    >
      <a
        href="/"
      >
        <img
          alt="companylogo"
          class="mb-[12vh]"
          src="/project2/bedrevLogoWhite.svg"
        />
      </a>
      <img
        alt="login"
        src="/project2/homepageArt.svg"
      />
    </div>
    <div
      class="flex-1 flex flex-col items-center font-whyte pt-[100px] sm:pt-[20vh] mx-[50px] sm:mx-[0px]"
    >
      <div
        class="w-full sm:min-w-[400px] sm:w-1/2"
      >
        <a
          href="/"
        >
          <img
            alt="companylogo"
            class="mb-8 sm:hidden"
            src="/project2/bedrevLogo.svg"
          />
        </a>
        <h3
          class="text-3xl text-primary tracking-wider font-light"
        >
          Registrer ny bruker
        </h3>
        <p
          class="pt-[20px] text-primary"
        >
          Fyll inn feltene for Ã¥ registrere deg
        </p>
        <form
          class="w-full mt-[20px] flex flex-col gap-4"
        >
          <div
            class="flex flex-1 h-8 flex-col"
          >
            <div
              class="flex"
            >
              <label
                class="font-medium mb-1 text-md flex justify-start w-full"
                for="radix-:r3:"
              >
                E-post
              </label>
            </div>
            <input
              class="w-[95%] rounded-sm font-inter tracking-wide text-md rounded-lg p-2 h-10 text-primary placeholder-darker outline-primary border-darker z-50 py-2 border-solid overflow-visible border-[2px]"
              id="radix-:r3:"
              name="email"
              placeholder="Skriv inn din E-post"
              title=""
              type="text"
              value=""
            />
          </div>
          <div
            class="flex flex-1 h-8 flex-col"
          >
            <div
              class="flex"
            >
              <label
                class="font-medium mb-1 text-md flex justify-start w-full"
                for="radix-:r4:"
              >
                Passord
              </label>
            </div>
            <input
              class="w-[95%] rounded-sm font-inter tracking-wide text-md rounded-lg p-2 h-10 text-primary placeholder-darker outline-primary border-darker z-50 py-2 border-solid overflow-visible border-[2px]"
              id="radix-:r4:"
              name="password"
              placeholder="Skriv inn ditt passord"
              title=""
              type="password"
              value=""
            />
          </div>
          <div
            class="flex flex-1 h-8 flex-col"
          >
            <div
              class="flex"
            >
              <label
                class="font-medium mb-1 text-md flex justify-start w-full"
                for="radix-:r5:"
              >
                Gjenta passord
              </label>
            </div>
            <input
              class="w-[95%] rounded-sm font-inter tracking-wide text-md rounded-lg p-2 h-10 text-primary placeholder-darker outline-primary border-darker z-50 py-2 border-solid overflow-visible border-[2px]"
              id="radix-:r5:"
              name="password"
              placeholder="Skriv inn ditt passord"
              title=""
              type="password"
              value=""
            />
          </div>
          <div
            class="mt-2 w-full"
          >
            <button
              class="bg-white text-black hover:bg-white hover:text-black rounded-sm border-black border-2 transition-all py-2 w-full text-sm px-undefined group transition hover:bg-primary border-solid"
              type="submit"
            >
              <div
                class="flex flex-row gap-2 justify-center pt-1 h-full px-6 align-middle"
              >
                Registrer deg
              </div>
            </button>
          </div>
          <a
            class="text-yellow self-end mt-1"
            href="/logg-inn"
          >
            Har du allerede bruker? Logg inn her
          </a>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
